### PR TITLE
tracing: Improve decoding of functions output

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Currently, while decoding result of function execution in traces, the following is done:
1. If function appears in ABIs of some contract seen during execution and identified by it's bytecode, ABI is taken from this contract bytecode.
2. Otherwise, ABI is fetched from openchain.xyz

This works fine for decoding inputs, however, openchain.xyz does not (and can not) tell anything about output types of a function.

Due to this, outputs for any calls to external contract accessed by an interface in forked environment are not decoded at the moment.

An example of this behavior is following test:
```solidity
pragma solidity 0.8.20;

import {Test} from "forge-std/Test.sol";

interface IERC20 {
    function balanceOf(address) external view returns (uint256);
}

contract TestTracing is Test {
    function test() public {
        vm.createSelectFork("mainnet");

        IERC20 usdc = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
        usdc.balanceOf(address(0));
    }
}
```

Traces for it will look like this:
```
  [15795] TestTracing::test()
    ├─ [0] VM::createSelectFork("mainnet")
    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000000
    ├─ [9815] 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48::balanceOf(0x0000000000000000000000000000000000000000) [staticcall]
    │   ├─ [2529] 0xa2327a938Febf5FEC13baCFb16Ae10EcBc4cbDCF::balanceOf(0x0000000000000000000000000000000000000000) [delegatecall]
    │   │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000000
    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000000
    └─ ← ()
```
However, it should be decoded into this:
```
  [15795] TestTracing::test()
    ├─ [0] VM::createSelectFork("mainnet")
    │   └─ ← 0
    ├─ [9815] 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48::balanceOf(0x0000000000000000000000000000000000000000) [staticcall]
    │   ├─ [2529] 0xa2327a938Febf5FEC13baCFb16Ae10EcBc4cbDCF::balanceOf(0x0000000000000000000000000000000000000000) [delegatecall]
    │   │   └─ ← 0
    │   └─ ← 0
    └─ ← ()
```

And this gets much uglier when working with contracts returning tuples/addresses/bytes
## Solution

There is already `with_events` function which is used for importing events from known contracts when constructing a `CallTraceDecoder`, so I've just added similar logic for functions as well
https://github.com/foundry-rs/foundry/blob/fdad9fb0dde45d3476fc5d1fe6f40e8dc7c17caa/crates/forge/bin/cmd/test/mod.rs#L692-L695

### CODE REVIEW

File: crates/forge/bin/cmd/test/mod.rs [hunk 1]
```diff
@@ -692,6 +693,7 @@ async fn test(
             let mut builder = CallTraceDecoderBuilder::new()
                 .with_labels(result.labeled_addresses.iter().map(|(a, s)| (*a, s.clone())))
                 .with_events(local_identifier.events().cloned())
+                .with_functions(local_identifier.functions().cloned())
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:696]: Can also add a helper function for local identifier that calls these 2 other methods

**Review by DaniPopes — CHANGES_REQUESTED** [review]
Can you also add a test with the example contract? I don't believe we have trace-specific tests, so a repro with this PR's number is OK. I think we'll want to separate those tho.

**Comment by DaniPopes** [comment]
Also cc @onbjerg @Evalir tracing change

**Comment by DaniPopes** [comment]
@klkvr I meant repro tests in crates/forge/tests/it/repros.rs

**Comment by klkvr** [comment]
Hey @DaniPopes I am not sure if it makes sense to add this as a repro test
1. It seems to me that it isn't possible to get `known_contracts` object and configure trace identifier with it while not changing tests too much
2. Even if it is possible, the test will not really make a lot of sense. I think test should check that 
    > traces are decoded correctly when testing through CLI

     rather than 

    >traces are decoded correctly when known contract abis are included into scope of call trace decoder

    The former is what this PR fixes, and the latter is _how_ it does that, and this logic already worked before this PR

**Review by mattsse — APPROVED** [review]
this is great!


the integration test is appropriate

**Pull request merged**